### ISSUE TITLE
BaseConfig should support array values with dot syntax

### ISSUE THREAD

**Issue by lonnieezell** [codeigniter4/CodeIgniter4#454]
Currently, I don't believe it loads those correctly. Check loading database group settings for a specific example.

**Comment by EpicKris** [comment]
I ran into the same issue when created a basic mail library for my application, seems to be because loading using dot syntax isn't recursive and only looks a few levels deep?

**Comment by lonnieezell** [comment]
That's what I thought it was, but I just realized that it seems to work fine for database config options, which is very similar to the mail options I've been working on. So this will probably take a little more investigation. :)

**Comment by DavisMiculis** [comment]
Hello, just a note on this same topic.

I am using current latest code (#df60c1b2) and I cannot load database config from .env file.
I've only tested it on mingrations.

.env:
```
database.default.hostname = 'localhost'
database.default.database = 'abc'
database.default.username = ''
database.default.password = ''
database.default.DBDriver = 'Postgre'
```

Database.php:
```
	public $default = [
		...
		'hostname'     => 'localhost',
		'username'     => '',
		'password'     => '',
		'database'     => 'def',
		'DBDriver'     => 'MySQLi',
		...
	];
```

Migration tries to connect to a mysql database and fails. It doesn't use the postgres credentials I specified in .env file.

**Comment by lonnieezell** [comment]
@DavisMiculis Where is your .env file? It sounds like it's not loading it. It should be under the root directory, not `application` directory where it used to live. I just stepped through a couple of examples on the database and it's reading settings correctly.

Additionally, it seems to be handling array values with dot syntax just fine, though I need to do some more tests to see why it wasn't working with my mail stuff.

**Comment by lonnieezell** [comment]
Ok, looks like my mail stuff wasn't working simply because it was deeper than the database config file is. So, the solution is either to simplify the mail config file or make BaseConfig recursive to deal with any depth - and that's probably the best solution.

**Comment by DavisMiculis** [comment]
Didn't even notice it was moved. Ok, so I moved .env file to root directory.
.env file still contains config for Postgres, while Config/Database.php has the default Mysql 

Here's something interesting:
If I cd into /public and run `php index.php migrate` it uses Postgres.
If I cd into root and run `php ci.php migrate:refresh` it uses Mysql.

But it still throws me errors:
```
davis@MacBook-Pro:~[codeigniter4]/public$ php index.php migrate

An uncaught Exception was encountered

Type:        Error
Message:     Call to undefined function CodeIgniter\Database\Postgre\pg_connect()
Filename:    [codeigniter4]/system/Database/Postgre/Connection.php
Line Number: 97
```
```
davis@MacBook-Pro:~[codeigniter4]$ php ci.php migrate:refresh

CodeIgniter CLI Tool - Version 4.0-dev - Server-Time: 2017-03-31 01:25:23am


An uncaught Exception was encountered

Type:        ErrorException
Message:     mysqli::real_connect(): (HY000/2002): No such file or directory
Filename:    [codeigniter4]/system/Database/MySQLi/Connection.php
Line Number: 183
```

**Comment by jim-parry** [comment]
PR 618 has unit tests for Config, including dot array syntax. 
Does it resolve this issue?

**Comment by lonnieezell** [comment]
@jim-parry IIRC, the issue is that it needs to handle it recursively. I believe it does it one or two levels deep only currently.

The other option is that we ditch the automatic setting of values since it is fairly magic, and provide a custom env() function to read values from the three places dotenv sets them. I've been thinking about that lately, but not sure what everyone else thinks....

**Comment by ujjaldutta** [comment]
can you allow to put function in database.php to use like 'database' => env('database.default.username', 'testdb'),
Its now not allowing me to use env file and function

**Comment by lonnieezell** [comment]
@ujjaldutta I use the env file for database config all of the time. Here's an example from my .env file:

```
### MYSQL LOCAL
database.default.hostname = localhost
database.default.database = ci4
database.default.username = root
database.default.password = root
database.default.DBDriver = MySQLi
database.default.DBPrefix = db_
```

And the db config is all default. 

However, there is an `env` method in the Common.php file already if you want to use that instead.

**Comment by lonnieezell** [comment]
@jim-parry What are your thoughts on removing the magic in the BaseConfig file altogether and just using the `env()` method?

**Comment by ujjaldutta** [comment]
Hi,
Can anyone tell me why in doc the function is $routes->resources . I have checked the function is $routes->resource not resources().
Please tell me if I am doing any mistake.
Thanks.

**Comment by makshh** [comment]
@ujjaldutta 

```css
@media only screen and (min-width: 60em) {
  .md-content {
    margin-right: 0;
  }
}
```

**Comment by lonnieezell** [comment]
@ujjaldutta Probably a typo.... and one that doesn't have anything to do with this issue. :)

**Comment by lonnieezell** [comment]
@jim-parry Since I haven't heard back an opinion just yet, I'm going to assume that means you're cool with scrapping the magical nature of BaseConfig. :) 

Seriously, though, I'm leaning very heavily toward doing that because it's a magic box that is hard to debug when it doesn't work quite right. using `env('DB_USER')` in the config files is:

- a) more explicit
- b) simpler to understand when you take the project over

If you don't have any issues, or if I don't hear from you in the next few days, I'll go ahead and remove that feature.

**Comment by jim-parry** [comment]
@lonnieezell Sorry - I thought I had commented, but I agree - it isn't here :-/
I have no problem simplifying the config, or things that rely on "magic" :)

**Comment by lonnieezell** [comment]
No worries. The main problem I have isn't that it's magic, it's that it's unclear how it's happening, and a PITA to debug when you're trying to get the right config value in the .env file. 

Since you're good either way, I'll go ahead and simplify that aspect. Won't lose any power and is clearer to boot.

**Comment by lonnieezell** [comment]
Or maybe not. I remember why I did that now. You can't implement expressions as default values. This is the one time that using a class for a config value really gets in the way, as compared to a simple array. :( 

Guess it will stay. The alternative is ugly.

**Comment by lonnieezell** [comment]
@ujjaldutta For your previous comment: no, we cannot allow use of `env()` to be assigned as the value because PHP won't allow you to use an expression as a default value. 

Instead, you can do it inside the constructor, though. 

```
public function __construct()
{
    parent::__construct();
    $this->database = env('database.default.username', 'testdb');
}
```

**Comment by jim-parry** [comment]
@lonnieezell Can we close this? anything else to fix or is this still a release blocker?

**Comment by lonnieezell** [comment]
I wouldn't say it's a release blocker, but is still a feature that would be nice to have, since dot array syntax works elsewhere. The main, thing, though, is that it should be updated to be recursive so it can handle more than 1 or 2 levels deep that it currently does.

**Comment by lonnieezell** [comment]
Fixed with #2082
